### DISCUSSION TITLE
STM32F4xx   25Mhz Flash emulation

### DISCUSSION THREAD

**Discussion by samitray** [grblHAL/core#503] [General]
For Stm32F4xx EEPROM_ENABLE 0 in my_machine.h on a custom stm32F446RE board with 25MHz crystal no eeprom chip, (EEPROM arrives after a week) Using Release F446 25MHz configuration, getting build errors because _EEPROM_Emul_Start and _EEPROM_Emul_Sector remain undefined,
So I change to Release F446 12MHz and now build completes ok because defines are found in the linker script, but in this 
configuration there is no space for bootloader, since my knowledge of microprocessors is pretty limited and before I nuke the 
bootloader on the chip... i presume that, since I use stlink to flash the program, i do not require the bootloader and all will be ok
if I change the HSE_VALUE value to 25M in the Project Settings / Paths and Symbols and flash..... or may be there is more to it....

**Comment by terjeio** [comment]
> i presume that, since I use stlink to flash the program, i do not require the bootloader and all will be ok

> if I change the HSE_VALUE value to 25M in the Project Settings / Paths and Symbols and flash.....

Correct.

**Reply by samitray** [reply]
The flashing works just by changing HSE_VALUE value, grblHAL starts up and jog commands are working ok. 
However if i try to change settings, it fails. I tired both from Candle as well as Serial port emulator
$1=255
[MSG:2]
[MSG:1]
[MSG:0]
ok
[MSG:Warning: Settings Write failed!]
In my_machine.h I have  #define EEPROM_ENABLE 0 uncommented.... will try and comment out that line and see how it goes.

**Reply by samitray** [reply]
Commented  #define EEPROM_ENABLE 0 , still similar problem...

**Reply by terjeio** [reply]
Oops, wrong linker script, change to `STM32F446RETX_FLASH.ld` or `STM32F446RETX_NO_BL_FLASH.ld`.

![image](https://github.com/grblHAL/core/assets/20260062/3670fe1b-c9b1-48bd-ae55-42bfb4d88ab0)

**Reply by samitray** [reply]
Issue reported above is with STM32F446RETX_NO_BL_FLASH.ld
![Screenshot from 2024-04-30 10-45-11](https://github.com/grblHAL/core/assets/4198282/34637388-d03b-41b7-afe8-1262608b32df)

will try with  STM32F446RETX_FLASH.ld and report back, need to modify / remove some defined symbols as compile is throwing a bunch of warnings.

**Reply by samitray** [reply]
Some more headaches... i use  stlink-org tool on linux to flash the STM32F4xx, prior to CubeIDE upgrade, my stlink worked fine and was able to flash from inside CubeIDE, but that has stopped working after cubeIDE upgrade, so I generate bin file in cubeIDE and then flash from command line using stlink-org.... Flashing works since I have flashed STM32F1xx grblHAL several times, However stlink-org command line need to have starting flash address. In case of STM32F446RETX_NO_BL_FLASH.ld the flash map starting address was 0x8000000 and emulation was after this address so everything was easy, Now in case of STM32F446RETX_FLASH.ld the emulation section is before and there is also a BOOT section, so I am confused as to what starting address to give on stlink-org for flashing... will try with 0x8000000 and report back...

**Reply by terjeio** [reply]
> will try with 0x8000000 and report back...

Correct, the address is [here](https://github.com/grblHAL/STM32F4xx/blob/555ebd03991a5213a71fc09d324201b9ea68b03c/STM32F446RETX_NO_BL_FLASH.ld#L40).
FYI a boot section is not written to, it assumes a preflashed bootloader that gets the start address from the vector table (at FLASH address).

**Reply by samitray** [reply]
Ok....I changed configuration to STM32F446RETX_FLASH.ld and the compile threw up a bunch of warnings possibly due to the symbols defined in project settings. So I made a copy of STM32F446RETX_NO_BL_FLASH.ld and replaced the memory map like below
`MEMORY
{
  RAM    (xrw)    : ORIGIN = 0x20000000,   LENGTH = 128K
  FLASH    (rx)    : ORIGIN = 0x8000000,   LENGTH = 512K - 128K
  EEPROM_EMUL(xrw)      : ORIGIN = 0x8060000,   LENGTH = 128K
}

_FLASH_VectorTable = ORIGIN(FLASH);
_EEPROM_Emul_Start = ORIGIN(EEPROM_EMUL);
_EEPROM_Emul_Sector = 11;`

to 
`MEMORY
{
  RAM    (xrw)    : ORIGIN = 0x20000000,   LENGTH = 128K
  BOOT_FLASH    (rx)    : ORIGIN = 0x8000000,   LENGTH = 16K
  EEPROM_EMUL(xrw)      : ORIGIN = 0x8004000,   LENGTH = 16K
  FLASH    (rx)    : ORIGIN = 0x8008000,   LENGTH = 512K - 16K - 16K
}

_EEPROM_Emul_Start = ORIGIN(EEPROM_EMUL);
_EEPROM_Emul_Sector = 1;`

and modified the isr_vector section to BOOT_FLASH
Then flashed the bin file at location 0x8008000 and connected to grblHAL and changed settings worked kind of... earlier it was failing, now it stores the values but prints a bunch of MSG codes

**Reply by samitray** [reply]
On settings change grblHAL is alway sending 3 MSG codes before ok.... I looked at documentation $EA , $ES, but nothing for MSG codes.... Candle expects ok response, but because of the message codes, Candle thinks that settings change has failed....
$1=26
[MSG:2]
[MSG:1]
[MSG:0]
ok

**Reply by terjeio** [reply]
> ... but prints a bunch of MSG codes

These are stray debug messages, removed in recent commits.

**Reply by samitray** [reply]
Any tips on how to locate the MSG codes... I looked at nvs_buffer.c line 372 where there is a retry loop for flash writing of settings, which prints the failed msg if the loop fails 4 times....so nothing to indicate how the MSG codes are generated....

**Reply by samitray** [reply]
ok, let me try to refresh the code....

**Comment by samitray** [comment]
@terjeio 
In STM32F446RETX_NO_BL_FLASH.ld related to configuration Release F446 12Mhz , After changing 
_EEPROM_Emul_Sector = 11;
to
 _EEPROM_Emul_Sector = 7;
as per this table in Ref manual
![Screenshot from 2024-04-30 12-18-13](https://github.com/grblHAL/core/assets/4198282/72c5584d-5892-47dc-bcda-35a57f178130)

The settings change fail message goes away

**Comment by samitray** [comment]
@terjeio  Thanks for the commit
